### PR TITLE
Add filter functionality to Try.

### PR DESCRIPTION
**Pull request by whirlwin**
Add filter functionality to Try.

Not sure what the the Failure should contain in case of failure, so I let it contain a generic `RuntimeException`.

### CODE REVIEW

File: src/main/java/no/finntech/lambdacompanion/Success.java [hunk 1]
```diff
@@ -32,6 +34,16 @@ public Success(T t) {
     }
 
     @Override
+    public Try<T> filter(final Predicate<T> predicate) {
+        Objects.requireNonNull(predicate);
+        if (predicate.test(t)) {
+            return this;
+        } else {
+            return new Failure<>(new RuntimeException());
```
Comment (nicolasyanncouturier) [inline src/main/java/no/finntech/lambdacompanion/Success.java]: A bit unsure about leaving it as a raw `RuntimeException` without even a message as it makes it impossible to know what happened.
What about a `NoSuchElementException("No value present")` just like in java8's `Optional#get` class? Maybe even with a more self explanatory message like f.ex. `"No success present after filtering"` or something like that?

**Comment by eirslett** [comment]
> Not sure what the the Failure should contain in case of failure

There should be a `Supplier<? extends Exception> onFailure` argument, forcing developers to provide proper exceptions.

**Comment by whirlwin** [comment]
@eirslett I think in most cases when using `filter`, you are not interested in the exception itself, but rather if filter yields a match or not. Providing an exception would IMO just clutter the code more than necessary.

**Comment by nicolasyanncouturier** [comment]
I originally was a bit puzzled by the idea that filtering results in a failure. The semantic of filtering is only to restrict further operations to a subset of the original value, i.e. `Stream#filter` or `Optional#filter` do not turn into a failure of some sort in case there is no match; it just does not do anything.

Similarly, in Scala, `Either.RightProjection#filter` (http://www.scala-lang.org/api/2.9.3/scala/Either$$RightProjection.html) does not turn a `Right` into a `Left` upon predicate failure. Rather, `Either.RightProjection#filter` returns an `Option[Either[L, R]]`. And so does lambda-companion's `Either`.

My opinion is that `Try#filter` should work in a similar manner by having the following signature: `public Optional<Try<T>> filter(final Predicate<T> predicate)`. It makes more sense both semantically and for the sake of coherence with lamba-companion's `Either`.

**Comment by whirlwin** [comment]
Maybe it helps a bit if I try to explain what the challenges I have. This is what I'm doing now which is kind of nasty:

``` java
Try<User> user = Try.of(() -> userService.getUser(1))
        .flatMap(maybeUser -> maybeUser.map(u -> Try.of(() -> u)).orElse(Try.failure(new NoSuchElementException()));
```

And this is what I'm hoping to do somehow (un-tested):

``` java
Try<User> user = Try.of(() -> userService.getUser(1))
        .filter(Optional::isPresent);
```

**Comment by eirslett** [comment]
Given another example, a dating app: You want to contact a user, and there can be two states other than success:

1) "This user doesn't exist"
vs
2) "The user is too young for you to make contact",

which are two different error conditions, and should arguably be shown to the user as different messages. We shouldn't lie to the user about which one of them is the problem.

A developer could be tempted and use something like `.filter(userIsSimilarAge)` (because it's easy and obvious), but then end up with a NoSuchElementException, which the server might mistakenly display as "This user doesn't exist". So in the `.filter` method we lose information. Which might be fine if it weren't for this project being lambda-companion - I suppose the reason lambda-companion exists in the first place, is to use the compiler to force people to handle errors correctly?

Doing like @nicolasyanncouturier suggests (Option), it could be expressed as something like

```
getUser(1)
  .filter(userIsSimilarAge).getOrElse(() -> new UserTooYoungException())
  .fold(handleErrors, displayUser)
```

**Comment by whirlwin** [comment]
Yes. That works, but (I think) `getOrElse` would have to return a `Try` in such case. So the code would be as follows:

``` java
getUser(1)
  .filter(userIsSimilarAge).getOrElse(() -> Try.failure(new UserTooYoungException()))
  .recover(handleErrors, displayUser)
```

I have updated the branch accordingly.

**Comment by nicolasyanncouturier** [comment]
LGTM

By the way, _if you do not care about the exception_ (i.e. fallback is to return a "dummy user") you could also always do a 

```
getUser(1)
    .toOptional()
    .flatMap(Function.identity())
    .orElseGet(this::anonymousUser)
```

**Pull request merged**